<a href="https://colab.research.google.com/github/Omkar210/Statistics-and-ML/blob/main/Day10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
# import warnings
# warnings.filterwarnings("ignore")
# pd.set_option('display.max_columns', None)
from numpy import random
from collections import Counter
from numpy.linalg import inv
from numpy.linalg import eig
import matplotlib
from matplotlib import pyplot as plt
import seaborn as sns
import pylab
from pylab import legend
from pylab import plot, show, title, xlabel, ylabel
import scipy
from scipy import stats
from scipy.stats import binom,poisson,norm,t,expon,f
from sklearn.model_selection import train_test_split
import statsmodels
from statsmodels import stats
from statsmodels.stats import weightstats as ssw
import os
print(os.getcwd())
os.chdir('/content/drive/MyDrive/CDAC')
os.getcwd()

/content/drive/MyDrive/CDAC


'/content/drive/MyDrive/CDAC'

In [ ]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='faithful')
df.head()

,eruptions,waiting
0,3.600,79
1,1.800,54
2,3.333,74
3,2.283,62
4,4.533,85


In [ ]:
erupt = df.eruptions

In [ ]:
xbar = np.mean(erupt)
n = len(erupt)
sd = np.std(erupt,ddof = 1)
print(xbar,n,sd)

3.4877830882352936 272 1.1413712511052077


# Tests on Categorical

### One sample Z-Test or One Sample T-Test

Z-test: only for continuous data, sample mean is > 30.

- Used to check whether the difference between the sample mean and claim value can be ignored or not.

- Used to check whether the difference between the sample mean and claim value is statistically significant or not

Step 1:
Hₒ : μ >= 3.7
Hₐ : μ < 3.7

Step 2:
CL = 95% = 0.95
CL = 5% = 0.05

Step 3:
Because Mean has to test with Claim mean.

Step 4:
Test Statistics , p-value

Step 5:


In [ ]:
ssw.ztest(erupt, value=3.7, alternative='smaller') # calculating the z-value and p-value respectively

(np.float64(-3.06646147570652), np.float64(0.0010830435959524984))

OR

In [ ]:
z = (xbar - 3.7)/(sd/n**0.5)
z


np.float64(-3.066461475706521)

In [ ]:
# 5% value on x-axis is called Zc
norm.ppf(0.05)

np.float64(-1.6448536269514729)

start point of 5% on standard normal curve is -1.6448536269514729

and z is -3.066461475706521 and p-value(0.0010830435959524984)

because the z-value is lesser than start point of 5% so we reject the NULL HYPOTHESIS.

Upto -1.645 is okay to be DO NOT REJECT.

---

There is 2 options to choose between Rejection or not to NULL HYPOTHESIS
1. Compare with z-value:

**Rejection will happen if z-value is more than 5%**

z-value is value on x-axis, if it's magnitude is more than Zc then we REJECT

2. Compare with p-value:

p-value is percentage, if it's magnitude is less than LOS then REJECT.

---
Calculation:

p-value -> norm.cdf(z)

In [ ]:
# Claim mean is 3.53

p = ssw.ztest(erupt, value=3.53, alternative='smaller')

# OR

z = (xbar - 3.53)/(sd/n**0.5)
print(f'z-value = {z}, \nz-value(stats.model) = {p[0]}, \np-value(stats.model) = {p[1]}')

z-value = -0.6100198729369253, 
z-value(stats.model) = -0.6100198729369251, 
p-value(stats.model) = 0.27092432161445423


In [ ]:
# p-value
norm.cdf(z)

np.float64(0.2709243216144542)

In [ ]:
# Null Hypthesis : mean < 3.37

print(ssw.ztest(erupt, value=3.37, alternative='larger')) # larger becz right tail test

print((xbar-3.37)/(sd/n**0.5))  # z-value

print(norm.ppf(0.95))  # ppf will take the left side area

print(1-norm.cdf(1.7019251649638607))  # cdf will take the z-value for calculating the p-value.

(np.float64(1.7019251649638607), np.float64(0.044384698843295746))
1.7019251649638614
1.6448536269514722
0.04438469884329577


In [ ]:
'''
Q. min waiting time 75min
test stat?
p-value?
95% ub
'''

wait = df.waiting
xbar = np.mean(wait)
n = len(wait)
sd = np.std(wait,ddof=1)

z = (xbar-75)/(sd/n**0.5)

print(z) # test statistics
print(norm.cdf(z)) # p-value
print(norm.ppf(0.95))  # area of 95%
print(xbar + norm.ppf(0.95)*sd/n**0.5)  # 95% Upperbound

# OR

print(ssw.ztest(wait, value=75, alternative='smaller'))  # left tail test so 'smaller'

-4.977386527579323
3.222427049500585e-07
1.6448536269514722
72.25293858852092
(np.float64(-4.9773865275793225), np.float64(3.222427049500608e-07))


# 2 sample t-test

this test is used when the difference of means of two independent samples has to be compared against the claim value and established whether their difference is statistically significant or not.

Area of Usage:

Two Independant Sample

In [ ]:
'''
Travel in Blore is atleast 15 min more than Mumbai

Blore
Sample size 50
Avg time 45
Std Dev 10.5

Mumbai
Sample size 60
Avg time 34
Std Dev 5.6
'''

'\nTravel in Blore is atleast 15 min more than Mumbai\n\nBlore\nSample size 50\nAvg time 45\nStd Dev 10.5\n\nMumbai\nSample size 60\nAvg time 34\nStd Dev 5.6\n'

In [ ]:
df1 = df.iloc[20:60]
df2 = df.iloc[120:170]

df1.shape

(40, 2)

## mean of erupt 2 is atleast 0.12 more than the mean of erupt1

Hₒ : mean(erupt2) - mean(erupt1) >= 0.12

Ha : mean(erupt2) - mean(erupt1) < 0.12

In [ ]:
erupt1 = df1.eruptions
erupt2 = df2.eruptions

In [ ]:
ssw.ztest(erupt2,erupt1,value=0.12,alternative='smaller')

(np.float64(-0.3196536558502502), np.float64(0.3746154475542464))

In [ ]:
xbar = np.mean(erupt1)
n1 = len(erupt1)
sd1 = np.std(erupt1,ddof = 1)
n2 = len(erupt2)
sd2 = np.std(erupt2,ddof = 1)
Nr = ((np.mean(erupt2) - np.mean(erupt1)) - 0.12)
Dr = ((sd1**2/n1) + (sd2**2/n2))**0.5

z = Nr/Dr
z

np.float64(-0.31844451421709036)

In [ ]:
ssw.ztest(erupt2,erupt1,value=0,alternative='two-sided')

(np.float64(0.160968305082337), np.float64(0.8721183639481939))

In [ ]:
df = pd.read_csv('6-Churn_ModellingDataSet.csv')
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [ ]:
x_age = df.Age
xb_age = np.mean(x_age)
n = len(x_age)
sd1 = np.std(x_age,ddof=1)

x_exit = df.Exited
xb_exit = np.mean(x_exit)
n = len(x_exit)
sd2 = np.std(x_exit,ddof=1)

Nr = ((np.mean(erupt2) - np.mean(erupt1)) - 0.12)
Dr = ((sd1**2/n1) + (sd2**2/n2))**0.5

z = Nr/Dr
z

np.float64(-0.048100167566537295)